File: BHAVRP_Python/cujae/inf/citi/om/controller/utils/Tools.py
from typing import List
from ...problem.input.problem import Problem
from ...problem.input.customer import Customer
import random
import numpy

class Tools:
    
    @staticmethod
    def random_ordenate():
        random.shuffle(Problem.get_problem().depots)
        
        copy_cost_matrix = Problem.get_problem().cost_matrix.copy()
        total_depots = Problem.get_problem().get_total_depots()
        total_customers = Problem.get_problem().get_total_customers()
        
        for _ in range(total_depots):
            copy_cost_matrix = numpy.delete(copy_cost_matrix, total_customers, axis=1)
        
        copy_depots = Problem.get_problem().depots.copy()
        
        for j in range(len(copy_depots)):
            pos_depot = Problem.get_problem().find_pos_depot(copy_depots, Problem.get_problem().depots[j].id_depot)
            new_col = Problem.get_problem().cost_matrix[:, total_customers + pos_depot].reshape(-1, 1)
            copy_cost_matrix = numpy.hstack((copy_cost_matrix, new_col))
        
        for _ in range(total_depots):
            copy_cost_matrix = numpy.delete(copy_cost_matrix, total_customers, axis=0)
        
        for k in range(len(copy_depots)):
            pos_depot = Problem.get_problem().find_pos_depot(copy_depots, Problem.get_problem().depots[k].id_depot)
            new_row = Problem.get_problem().cost_matrix[total_customers + pos_depot, :].reshape(1, -1)
            copy_cost_matrix = numpy.vstack((copy_cost_matrix, new_row))
        
        Problem.get_problem().set_cost_matrix(copy_cost_matrix)
        
    @staticmethod
    def descendent_ordenate():
        copy_depots = Problem.get_problem().depots.copy()
        total_depots = Problem.get_problem().get_total_depots()
            
        for i in range(total_depots - 1):
            for j in range(total_depots - i - 1):
                if Problem.get_problem().get_total_capacity_by_depot(Problem.get_problem().depots[j + 1]) > Problem.get_problem().get_total_capacity_by_depot(Problem.get_problem().depots[j]):
                    Problem.get_problem().depots[j], Problem.get_problem().depots[j + 1] = Problem.get_problem().depots[j + 1], Problem.get_problem().depots[j]
            
        copy_cost_matrix = Problem.get_problem().cost_matrix.copy()
        total_customers = Problem.get_problem().get_total_customers()
            
        for _ in range(total_depots):
            copy_cost_matrix = numpy.delete(copy_cost_matrix, total_customers, axis=1)
            
        for j in range(len(copy_depots)):
            pos_depot = Problem.get_problem().find_pos_depot(copy_depots, Problem.get_problem().depots[j].id_depot)
            new_col = Problem.get_problem().cost_matrix[:, total_customers + pos_depot].reshape(-1, 1)
            copy_cost_matrix = numpy.hstack((copy_cost_matrix, new_col))
                
        for _ in range(total_depots):
            copy_cost_matrix = numpy.delete(copy_cost_matrix, total_customers, axis=0)
            
        for k in range(len(copy_depots)):
            pos_depot = Problem.get_problem().find_pos_depot(copy_depots, Problem.get_problem().depots[k].id_depot)
            new_row = Problem.get_problem().cost_matrix[total_customers + pos_depot, :].reshape(1, -1)
            copy_cost_matrix = numpy.vstack((copy_cost_matrix, new_row))
            
        Problem.get_problem().set_cost_matrix(copy_cost_matrix)

    @staticmethod
    def ascendent_ordenate():
        copy_depots = Problem.get_problem().depots.copy()
        total_depots = Problem.get_problem().get_total_depots()
            
        for i in range(total_depots - 1):
            for j in range(total_depots - i - 1):
                if Problem.get_problem().get_total_capacity_by_depot(Problem.get_problem().depots[j + 1]) < Problem.get_problem().get_total_capacity_by_depot(Problem.get_problem().depots[j]):
                    Problem.get_problem().depots[j], Problem.get_problem().depots[j + 1] = Problem.get_problem().depots[j + 1], Problem.get_problem().depots[j]
            
        copy_cost_matrix = Problem.get_problem().cost_matrix.copy()
        total_customers = Problem.get_problem().get_total_customers()
            
        for _ in range(total_depots):
            copy_cost_matrix = numpy.delete(copy_cost_matrix, total_customers, axis=1)
            
        for j in range(len(copy_depots)):
            pos_depot = Problem.get_problem().find_pos_depot(copy_depots, Problem.get_problem().depots[j].id_depot)
            new_col = Problem.get_problem().cost_matrix[:, total_customers + pos_depot].reshape(-1, 1)
            copy_cost_matrix = numpy.hstack((copy_cost_matrix, new_col))
                
        for _ in range(total_depots):
            copy_cost_matrix = numpy.delete(copy_cost_matrix, total_customers, axis=0)
            
        for k in range(len(copy_depots)):
            pos_depot = Problem.get_problem().find_pos_depot(copy_depots, Problem.get_problem().depots[k].id_depot)
            new_row = Problem.get_problem().cost_matrix[total_customers + pos_depot, :].reshape(1, -1)
            copy_cost_matrix = numpy.vstack((copy_cost_matrix, new_row))
            
        Problem.get_problem().set_cost_matrix(copy_cost_matrix)

    @staticmethod
    def random_ordenate_customers(customers: List['Customer']):
        random.shuffle(customers)

    @staticmethod
    def descendent_ordenate_customers(customers: List['Customer']):
        customers.sort(key=lambda c: Problem.get_problem().get_request_by_id_customer(c.id_customer), reverse=True)

    @staticmethod
    def ascendent_ordenate_customers(customers: List['Customer']):
        customers.sort(key=lambda c: Problem.get_problem().get_request_by_id_customer(c.id_customer))

    @staticmethod
    def truncate_double(number: float, decimal_place: int) -> float:
        return round(number, decimal_place)

File: BHAVRP_Java/BHAVRPv1.0/src/cujae/inf/citi/om/heuristic/controller/Tools.java
package cujae.inf.citi.om.heuristic.controller;
import cujae.inf.citi.om.problem.input.*;

import libmatrix.cujae.inf.citi.om.matrix.NumericMatrix;

import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.Random;

public abstract class Tools {
	
	public static void randomOrdenate(){
		Random random = new Random();
		int posRandom = -1;
		ArrayList<Depot> orderedDepots = new ArrayList<Depot>();
		ArrayList<Depot> copyDepots = new ArrayList<Depot>(InfoProblem.getProblem().getDepots());
		
		while(!InfoProblem.getProblem().getDepots().isEmpty())
		{
			posRandom = random.nextInt(InfoProblem.getProblem().getDepots().size());
			orderedDepots.add(InfoProblem.getProblem().getDepots().remove(posRandom));
		}
		
		for(int i = 0; i < orderedDepots.size(); i++)
			InfoProblem.getProblem().getDepots().add(orderedDepots.get(i));
		
		NumericMatrix copyCostMatrix = new NumericMatrix(InfoProblem.getProblem().getCostMatrix());
		int currentDepots = 0;
		int totalDepots = InfoProblem.getProblem().getTotalDepots();
		int totalCustomers = InfoProblem.getProblem().getTotalCustomers();
		
		while(currentDepots < totalDepots)
		{
			copyCostMatrix.deleteCol(totalCustomers);			
			currentDepots++;
		}
			
		int posDepot;
		
		for(int j = 0; j < copyDepots.size(); j++)
		{
			posDepot = InfoProblem.getProblem().findPosDepot(copyDepots, InfoProblem.getProblem().getDepots().get(j).getIDDepot());
			copyCostMatrix.addCol((totalCustomers + j), InfoProblem.getProblem().getCostMatrix().getCol(totalCustomers + posDepot));
		}	
		
		currentDepots = 0;
		
		while(currentDepots < totalDepots)
		{
			copyCostMatrix.deleteRow(totalCustomers);
			currentDepots++;
		}
		
		for(int k = 0; k < copyDepots.size(); k++)
		{
			posDepot = InfoProblem.getProblem().findPosDepot(copyDepots, InfoProblem.getProblem().getDepots().get(k).getIDDepot());
			copyCostMatrix.addRow((totalCustomers + k), InfoProblem.getProblem().getCostMatrix().getRow(totalCustomers + posDepot));
		}	
		
		InfoProblem.getProblem().setCostMatrix(copyCostMatrix);
	}
	
	public static void descendentOrdenate(){
		Depot depot = new Depot();
		ArrayList<Depot> copyDepots = new ArrayList<Depot>(InfoProblem.getProblem().getDepots());
		int totalDepots = InfoProblem.getProblem().getTotalDepots();
		
		for(int i = 0; i < (totalDepots - 1); i++)
		{
			for(int j = 0; j < (totalDepots - i - 1); j++)
			{
				if(InfoProblem.getProblem().getTotalCapacityByDepot(InfoProblem.getProblem().getDepots().get(j + 1)) > InfoProblem.getProblem().getTotalCapacityByDepot(InfoProblem.getProblem().getDepots().get(j)))
				{
					depot = InfoProblem.getProblem().getDepots().get(j + 1);
					InfoProblem.getProblem().getDepots().set((j + 1), InfoProblem.getProblem().getDepots().get(j));
					InfoProblem.getProblem().getDepots().set(j, depot);
				}
			}
		}
		
		NumericMatrix copyCostMatrix = new NumericMatrix(InfoProblem.getProblem().getCostMatrix());
		int currentDepots = 0;
		int totalCustomers = InfoProblem.getProblem().getTotalCustomers();
		
		while(currentDepots < totalDepots)
		{
			copyCostMatrix.deleteCol(totalCustomers);
			currentDepots++;
		}
			
		int posDepot;
		
		for(int j = 0; j < copyDepots.size(); j++)
		{
			posDepot = InfoProblem.getProblem().findPosDepot(copyDepots, InfoProblem.getProblem().getDepots().get(j).getIDDepot());
			copyCostMatrix.addCol((totalCustomers + j), InfoProblem.getProblem().getCostMatrix().getCol(totalCustomers + posDepot));
		}	
		
		currentDepots = 0;
		
		while(currentDepots < totalDepots)
		{
			copyCostMatrix.deleteRow(totalCustomers);
			currentDepots++;
		}
		
		for(int k = 0; k < copyDepots.size(); k++)
		{
			posDepot = InfoProblem.getProblem().findPosDepot(copyDepots, InfoProblem.getProblem().getDepots().get(k).getIDDepot());
			copyCostMatrix.addRow((totalCustomers + k), InfoProblem.getProblem().getCostMatrix().getRow(totalCustomers + posDepot));
		}	
		
		InfoProblem.getProblem().setCostMatrix(copyCostMatrix);
	}

	public static void ascendentOrdenate(){
		Depot depot = new Depot();
		ArrayList<Depot> copyDepots = new ArrayList<Depot>(InfoProblem.getProblem().getDepots());
		int totalDepots = InfoProblem.getProblem().getTotalDepots();
		
		for(int i = 0; i < (totalDepots - 1); i++)
		{
			for(int j = 0; j < (totalDepots - i - 1); j++)
			{
				if(InfoProblem.getProblem().getTotalCapacityByDepot(InfoProblem.getProblem().getDepots().get(j + 1)) < InfoProblem.getProblem().getTotalCapacityByDepot(InfoProblem.getProblem().getDepots().get(j)))
				{
					depot = InfoProblem.getProblem().getDepots().get(j + 1);
					InfoProblem.getProblem().getDepots().set((j + 1), InfoProblem.getProblem().getDepots().get(j));
					InfoProblem.getProblem().getDepots().set(j, depot);
				}
			}
		}
		
		NumericMatrix copyCostMatrix = new NumericMatrix(InfoProblem.getProblem().getCostMatrix());
		int currentDepots = 0;
		int totalCustomers = InfoProblem.getProblem().getTotalCustomers();
		
		while(currentDepots < totalDepots)
		{
			copyCostMatrix.deleteCol(totalCustomers);
			currentDepots++;
		}
			
		int posDepot;
		
		for(int j = 0; j < copyDepots.size(); j++)
		{
			posDepot = InfoProblem.getProblem().findPosDepot(copyDepots, InfoProblem.getProblem().getDepots().get(j).getIDDepot());
			copyCostMatrix.addCol((totalCustomers + j), InfoProblem.getProblem().getCostMatrix().getCol(totalCustomers + posDepot));
		}	
		
		currentDepots = 0;
		
		while(currentDepots < totalDepots)
		{
			copyCostMatrix.deleteRow(totalCustomers);
			currentDepots++;
		}
		
		for(int k = 0; k < copyDepots.size(); k++)
		{
			posDepot = InfoProblem.getProblem().findPosDepot(copyDepots, InfoProblem.getProblem().getDepots().get(k).getIDDepot());
			copyCostMatrix.addRow((totalCustomers + k), InfoProblem.getProblem().getCostMatrix().getRow(totalCustomers + posDepot));
		}	
		
		InfoProblem.getProblem().setCostMatrix(copyCostMatrix);
	}

	public static void randomOrdenate(ArrayList<Customer> customers){
		Random random = new Random();
		int posRandom = -1;
		ArrayList<Customer> orderedCustomers = new ArrayList<Customer>();
		
		while(!customers.isEmpty())
		{
			posRandom = random.nextInt(customers.size());
			orderedCustomers.add(customers.remove(posRandom));
		}
		
		for(int i = 0; i < orderedCustomers.size(); i++)
			customers.add(orderedCustomers.get(i));
		
		/* no necesito actualizar la matriz*/
	}

	public static void descendentOrdenate(ArrayList<Customer> customers){
		Customer customer = new Customer();
		int totalCustomers = customers.size();
		
		for(int i = 0; i < (totalCustomers - 1); i++)
		{
			for(int j = 0; j < (totalCustomers - i - 1); j++)
			{
				if(InfoProblem.getProblem().getRequestByIDCustomer(customers.get(j + 1).getIDCustomer()) > InfoProblem.getProblem().getRequestByIDCustomer(customers.get(j).getIDCustomer()))
				{
					customer = customers.get(j + 1);
					customers.set((j + 1), customers.get(j));
					customers.set(j, customer);
					
					/* no actualizo infoproblem*/
				}
			}
		}
	}

	public static void ascendentOrdenate(ArrayList<Customer> customers){
		Customer customer = new Customer();
		int totalCustomers = customers.size();
		
		for(int i = 0; i < (totalCustomers - 1); i++)
		{
			for(int j = 0; j < (totalCustomers - i - 1); j++)
			{
				if(InfoProblem.getProblem().getRequestByIDCustomer(customers.get(j + 1).getIDCustomer()) < InfoProblem.getProblem().getRequestByIDCustomer(customers.get(j).getIDCustomer()))
				{
					customer = customers.get(j + 1);
					customers.set(j+1, customers.get(j));
					customers.set(j, customer);
				}
			}
		}
	}
	
	public static double truncateDouble(double number, int decimalPlace) {
        double numberRound;
        BigDecimal bigDec = new BigDecimal(number);
        bigDec = bigDec.setScale(decimalPlace, BigDecimal.ROUND_DOWN);
        numberRound = bigDec.doubleValue();
        
        return numberRound;
    }
}
